rapid/workflow/queue.py

```diff
@@ -13,110 +13,60 @@
-                                if response.status_code == 201:
-                                    if 'X-Exclude-Resource'.lower() in response.headers:
-                                        client.sleep = True
-                                else:
-                                    logger.info("Client didn't respond right: {} returned {}".format(client.ip_address, response.status_code))
+    def setup_queue_handlers(self):
+        # type: () -> None
+        for handler in QueueHandlerConstants.queue_handler_classes:
+            self.queue_handlers.append(IOC.get_class_instance(handler))
 
-                                break  # Break, we sent work to the other client
-                            except Exception as exception:
-                                logger.error("Could not send work to worker: [{}]".format(str(exception)))
-                                self.action_instance_service.edit_action_instance(work_request.action_instance_id, {"status_id": StatusConstants.READY,
-                                                                                                                    "start_date": None,
-                                                                                                                    "assigned_to": None})
-                        except ConnectTimeout:
-                            # Should reset, this is a problem, server not there.
-                            self.action_instance_service.edit_action_instance(work_request.action_instance_id, {"status_id": StatusConstants.READY,
-                                                                                                                "start_date": None,
-                                                                                                                "assigned_to": None})
-                        except ReadTimeout as read_timeout:
-                            logger.error(read_timeout)
-                            logger.error(client.get_work_uri())
-                        except ConnectionError as error:
-                            logger.error(error)
-                            logger.error(client.get_work_uri())
-                        except Exception as exception:
-                            logger.error(exception)
-                            logger.error(client.get_work_uri())
+    def process_queue(self, clients):
+        for work_request in self.queue_service.get_current_work():
+            for queue_handler in self.queue_handlers:
+                if queue_handler.can_process_work_request(work_request):
+                    try:
+                        queue_handler.process_work_request(work_request, clients)
+                    except Exception as exception:
+                        logger.exception(exception)
+                        self.action_instance_service.edit_action_instance(work_request.action_instance_id, {'status_id': Constants.STATUS_FAILED})
+                    break
```

Comment:
means we give it to the first queue handler we find that says it can handle it. That keeps us from giving the same work to two handlers, if there are two handlers that say they can do a thing.